hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ProcessorTasklet.java

```diff
@@ -314,47 +290,46 @@ private void stateMachineStep() {
                 if (outbox.offerToEdgesAndSnapshot(currentBarrier)) {
                     progTracker.madeProgress();
                     if (currentBarrier.isTerminal()) {
-                        state = EMIT_DONE_ITEM;
+                        state = WAITING_FOR_SNAPSHOT_COMPLETED;
                     } else {
                         currentBarrier = null;
                         receivedBarriers.clear();
-                        pendingSnapshotId++;
-                        state = initialProcessingState();
+                        pendingSnapshotId1++;
+                        state = processingState();
                     }
                 }
-                progTracker.notDone();
                 return;
 
-            case COMPLETE:
-                progTracker.notDone();
-                // check ssContext to see if a barrier should be emitted
-                long currSnapshotId = ssContext.activeSnapshotId();
-                assert currSnapshotId + 1 == pendingSnapshotId || currSnapshotId == pendingSnapshotId
-                        : "Unexpected new snapshot id: " + currSnapshotId + ", expected was "
-                                + (pendingSnapshotId - 1) + " or " + pendingSnapshotId;
-                if (currSnapshotId == pendingSnapshotId) {
-                    if (outbox.hasUnfinishedItem()) {
-                        outbox.block();
-                    } else {
-                        outbox.unblock();
-                        state = SAVE_SNAPSHOT;
-                        currentBarrier = new SnapshotBarrier(currSnapshotId, ssContext.isTerminalSnapshot());
-                        progTracker.madeProgress();
-                        return;
-                    }
-                }
-                if (processor.complete()) {
+            case ON_SNAPSHOT_COMPLETED:
+            case ON_SNAPSHOT_COMPLETED_BEFORE_END:
+                if (processor.onSnapshotCompleted(ssContext.isLastPhase1Successful())) {
+                    pendingSnapshotId2++;
+                    ssContext.phase2DoneForTasklet();
                     progTracker.madeProgress();
-                    state = EMIT_DONE_ITEM;
+                    state = state == ON_SNAPSHOT_COMPLETED_BEFORE_END
+                            ? EMIT_DONE_ITEM
+                            : processingState();
+                }
+                return;
```

Comment:
This is very confusing code. I understand it now and it's probably perfectly correct but it's mad scientist code. It took me a while to understand that the handling of the two states actually differs and that `state = state ==` is also confusing at a first glance. (Btw. do my comments at least give you insight into how less intelligent people perceive the world?)

Reply:
ok, will refactor it to `if` :)

Reply:
The shape of this code is the unfortunate consequence of Java not having coroutines. In the processors, where state changes are data driven, we can usually survive using FP idioms, but this logic is too complex and needs a full state machine.